### ISSUE TITLE
Ignore entity type in migrations but not in query for shared type

### ISSUE THREAD

**Issue by chrfin** [aspnet/EntityFramework6#515]
I am building a web-application which is extendable via plugins which are loaded dynamically at runtime and where each plugin has his own entities and `DbContext` including migrations. This works fine as long as I do not have entities shared between my base context and a plugins context. 
But for example my `User` entity is defined in my base project and referenced/migrated in my common context.
If I then have a reference/relation in my plugins context to a user the plugins context complains about the missing migration for the `User` type. Until now I simply added a new migration to the plugin and "cleaned out" the `Up` and `Down` methods. But that does not work "in the field" when an old plugin should work with a newer version of the base class.

Now I tried the following:
`modelBuilder.Types<User>().Configure(c => c.Ignore());`
This did fix the context complaining about the migration, but I can't query the `User` type anymore.

What I could get working is create the empty migration at the beginning and then the following if the entity changes:
`modelBuilder.Types<User>().Configure(c => c.Ignore(u => u.Age));` where `Age` is the added property
This works and the "known" properties are still working in the plugin context. I could make this work "dynamically" if each plugin calls a "ignore properties newer than x"-method in the base class, but I think this is a little cumbersome and error-prone.

Is there a better solution to that problem or can EF easily be extended to support such a scenario?
I'm happy to do the extension myself if someone points me to the right place in the code.

### Steps to reproduce
You can use the following demo-solution to reproduce the problem: https://github.com/chrfin/EFMultiContextTest

### Further technical details
EF version: 6.2
Database Provider: EntityFramework.SqlServer
Operating system: Windows 10)
IDE: Visual Studio 2017 15.6

**Comment by chrfin** [comment]
I don't think so @chrysikos - EF handles my scenario very well in that matter and all CRUD operations work flawlessly. Only the migrations-part is not aware of the shared entities and tries to scaffold the schema change for every context (and how should it know to do otherwise if not told so - and the "if not told" part is the one I'm asking about).
I do not see how another layer would even help me solve my problem, as this would not change the basic DB context used for persistence, but feel free to show me how using my simple example - I'm happy to be proven wrong...

**Comment by chrysikos** [comment]
Can you tell me why PluginDbContextnew needs the User Entity?

**Comment by chrfin** [comment]
Because e.g. the plugin wan'ts to save which user create an item ;-)
That was just an example to show the problem in the simplest way possible. In "real life" there are several classes/entities which are reverenced/extended in plugins, but everything works aside from the problem highlighted here.

**Comment by chrysikos** [comment]
So your plugin needs only a small portion of a User not just entire reference of the DbSet<User>. Firstly, check Domain Driven Design and specifically Domain Events. If a Bounded Context, in your situation a plugin, need information from a different Bounded Context you must pass to it only (primitive types) to avoid coupling of these Contexts. If you don't understand that you must start study more about DDD. Here is some useful recourses:

Domain Events:
https://docs.microsoft.com/en-us/dotnet/standard/microservices-architecture/microservice-ddd-cqrs-patterns/domain-events-design-implementation

Eric Evans book:
https://www.amazon.com/Domain-Driven-Design-Tackling-Complexity-Software/dp/0321125215/ref=sr_1_1?s=books&ie=UTF8&qid=1522844020&sr=1-1&keywords=Eric+Evans

Entity Framework in the Enterprise by Julie Lerman
https://app.pluralsight.com/library/courses/entity-framework-enterprise-update/table-of-contents

Domain-Driven Design Fundamentals
by Steve Smith and Julie Lerman:
https://app.pluralsight.com/library/courses/domain-driven-design-fundamentals/table-of-contents

**Comment by chrfin** [comment]
Thanks for your additional input - I actually have Eric Evans book on my desk at the moment.
I also think my real application model is following DDD in may ways, but the coupling in this case is on purpose - I know that in this simple example case I could also just set a primitive `UserId` field to store the desired information. As already said this is just a simple broken down example to reproduce my problem.

My real application for example also has a shared `ContentItem` base type which most plugins extend to implement their own content items. By using TPT this works great as every plugin has their own "content item", but can share all the common logic (ordering, sorting, hierarchies, and many more...) via the base class. I know I could also solve this via a referenced base type and a new type for the extension (which would actually also mirror how TPT stores it in the DB), but that's not how I want to do it for several reasons. I have a very flexible and *working* solution here, which has a good separation of concerns and heavily follows the DRY principle.

To summarize: I do not want to get rid of the `User` reference inside my plugin entity (which I couldn't do anyhow, as the system is already "live" with such references), I just want to exclude it from migrations in the plugin to make the further development of these base-entities easier.

**Comment by ajcvickers** [comment]
@chrfin Unfortunately there is no great solution to automate this. The only way to do it is to edit the generated migrations so that shared entity types are handled by only one context type.

**Comment by chrfin** [comment]
Thanks for the info @ajcvickers - that's how I did it until now anyhow.
